rustfst-python/rustfst/fst.py

```diff
@@ -0,0 +1,458 @@
+import ctypes
+from rustfst.utils import (
+    lib,
+    check_ffi_error,
+)
+
+from rustfst.symbol_table import SymbolTable
+from rustfst.iterators import TrsIterator, StateIterator
+from rustfst.weight import weight_one
+
+class Fst(object):
+    """
+    Fst(ptr=None)
+      This class wraps a mutable FST and exposes all methods.
+      Args:
+        ptr: An optional pointer pointing to an existing Fst rust struct.
+    """
+
+    def __init__(self, ptr=None):
+        if ptr:
+            self._fst = ptr
+        else:
+            fst_ptr = ctypes.pointer(ctypes.c_void_p())
+            ret_code = lib.fst_new(ctypes.byref(fst_ptr))
+
+            err_msg = "Something went wrong when creating the Fst struct"
+            check_ffi_error(ret_code, err_msg)
+            self._fst = fst_ptr
+
+        # add shims for symbol tables (prevent early gc of the tables)
+        self._input_symbols = None
+        self._output_symbols = None
+
+    @property
+    def ptr(self):
+        return self._fst
+
+    def add_tr(self, state, tr):
+        """
+        add_tr(self, state, tr)
+            Adds a new tr to the FST and return self. Note the tr should be considered
+            consumed and is not safe to use it after.
+            Args:
+              state: The integer index of the source state.
+              tr: The tr to add.
+            Returns:
+              self.
+            Raises:
+              SnipsFstException: If State index out of range.
+            See also: `add_state`.
+        """
+        ret_code = lib.fst_add_tr(self._fst, ctypes.c_size_t(state),tr._ptr)
+        err_msg = "Error during `add_tr`"
+        check_ffi_error(ret_code, err_msg)
+
+        return self
+
+    def add_state(self):
+        """
+        add_state(self)
+            Adds a new state to the FST and returns the state ID.
+            Returns:
+              The integer index of the new state.
+            See also: `add_tr`, `set_start`, `set_final`.
+        """
+        state_id = ctypes.c_size_t()
+
+        ret_code = lib.fst_add_state(self._fst, ctypes.byref(state_id))
+        err_msg = "Error during `add_state`"
+        check_ffi_error(ret_code, err_msg)
+
+        return state_id.value
+
+    def trs(self, state):
```

Comment:
In all the python code, I believe it will be more user friendly to add the type of each parameter along with the return type : https://docs.python.org/3/library/typing.html